compiler/rustc_driver/src/lib.rs

```diff
@@ -1323,9 +1323,46 @@ pub fn init_env_logger(env: &str) {
     tracing::subscriber::set_global_default(subscriber).unwrap();
 }
 
+#[cfg(unix)]
+fn print_stack_trace(_: libc::c_int) {
+    unsafe {
+        static mut STACK_TRACE: [*mut libc::c_void; 256] = [std::ptr::null_mut(); 256];
+        let _depth = libc::backtrace(STACK_TRACE.as_mut_ptr(), 256);
```

Comment:
Libc's `backtrace` function does a naive walk of frame pointers. Rustc and LLVM are normally compiled with frame pointers omitted, meaning that only DWARF based unwinding is reliable.

https://www.gnu.org/software/libc/manual/html_node/Backtraces.html

> [...] frame pointer elimination will stop `backtrace` from interpreting the stack contents correctly.

In addition it is not safe to call it from asynchronous signals like SIGSEGV.

> [...] AS-Unsafe init heap dlopen plugin lock [...]

Reply:
Yeah I noticed inside the llvm code, using the backtrace c api was one of the ways they were generating the backtrace. I will try to add the other ways too.

Reply:
@bjorn3 [Here](https://github.com/rust-lang/llvm-project/blob/62a1ddde22c267249eda72184520a21ad2052f0b/llvm/lib/Support/Unix/Signals.inc#L558-L628) you can see, the glibc `backtrace` function is being used in llvm to print the backtrace. The alternative method being used is unwindBacktrace, but I think it is for ARM systems.

Reply:
@bjorn3 [Here](https://github.com/rust-lang/llvm-project/blob/62a1ddde22c267249eda72184520a21ad2052f0b/llvm/lib/Support/Unix/Signals.inc#L565), llvm uses the `backtrace` function to get the backtrace. If it fails it tries to use `_Unwind_Backtrace`.